src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,877 @@
+begin
+  dsimp [family_of_elements.is_amalgamation, yoneda_equiv, yoneda_lemma,
+         nat_trans_equiv_compatible_family],
+  split,
+  { rintro rfl,
+    intros Y f hf,
+    rw ← functor_to_types.naturality _ _ g,
+    change g.app (op Y) (f ≫ 𝟙 X) = g.app (op Y) f,
+    simp only [comp_id] },
+  { intro h,
+    ext Y ⟨f, hf⟩,
+    have : _ = x.app Y _ := h f hf,
+    rw [← this, ← functor_to_types.naturality _ _ g],
+    dsimp,
+    simp },
+end
+
+-- TODO: MOVE ME BEFORE PR
+lemma equiv.exists_unique_congr {α β : Type*} (p : β → Prop) (e : α ≃ β) :
+  (∃! (y : β), p y) ↔ ∃! (x : α), p (e x) :=
+begin
+  split,
+  { rintro ⟨b, hb₁, hb₂⟩,
+    exact ⟨e.symm b, by simpa using hb₁, λ x hx, by simp [←hb₂ (e x) hx]⟩ },
+  { rintro ⟨a, ha₁, ha₂⟩,
+    refine ⟨e a, ha₁, λ y hy, _⟩,
+    rw ← equiv.symm_apply_eq,
+    apply ha₂,
+    simpa using hy },
+end
+
+lemma yoneda_condition_iff_sheaf_condition :
+  is_sheaf_for P S ↔ yoneda_sheaf_condition P S :=
+begin
+  rw [is_sheaf_for, yoneda_sheaf_condition],
+  simp_rw [extension_iff_amalgamation],
+  rw equiv.forall_congr_left' nat_trans_equiv_compatible_family,
+  rw subtype.forall,
+  apply ball_congr,
+  intros x hx,
+  rw ← equiv.exists_unique_congr _ _,
+  simp,
+end
+
+lemma separated_for_and_exists_amalgamation_iff_sheaf_for :
+  is_separated_for P R ∧ (∀ (x : family_of_elements P R), x.compatible → ∃ t, x.is_amalgamation t) ↔ is_sheaf_for P R :=
+begin
+  rw [is_separated_for, ←forall_and_distrib],
+  apply forall_congr,
+  intro x,
+  split,
+  { intros z hx, exact exists_unique_of_exists_of_unique (z.2 hx) z.1 },
+  { intros h,
+    refine ⟨_, (exists_of_exists_unique ∘ h)⟩,
+    intros t₁ t₂ ht₁ ht₂,
+    apply (h _).unique ht₁ ht₂,
+    exact is_compatible_of_exists_amalgamation x ⟨_, ht₂⟩ }
+end
+
+lemma is_separated_for.is_sheaf_for (t : is_separated_for P R) :
+  (∀ (x : family_of_elements P R), x.compatible → ∃ t, x.is_amalgamation t) →
+  is_sheaf_for P R :=
+begin
+  rw ← separated_for_and_exists_amalgamation_iff_sheaf_for,
+  apply and.intro t,
+end
+
+/-- Get the amalgamation of the given compatible family, provided we have a sheaf. -/
+noncomputable def is_sheaf_for.amalgamate
+  (t : is_sheaf_for P R) (x : family_of_elements P R) (hx : x.compatible) :
+  P.obj (op X) :=
+classical.some (t x hx).exists
+
+lemma is_sheaf_for.is_amalgamation
+  (t : is_sheaf_for P R) {x : family_of_elements P R} (hx : x.compatible) :
+  x.is_amalgamation (t.amalgamate x hx) :=
+classical.some_spec (t x hx).exists
+
+@[simp]
+lemma is_sheaf_for.valid_glue
+  (t : is_sheaf_for P R) {x : family_of_elements P R} (hx : x.compatible) (f : Y ⟶ X) (Hf : R f) :
+  P.map f.op (t.amalgamate x hx) = x f Hf :=
+t.is_amalgamation hx f Hf
+
+/-- If `P` is a sheaf for `R`, it is separated for `R`. -/
+lemma is_sheaf_for.is_separated_for : is_sheaf_for P R → is_separated_for P R :=
+λ q, (separated_for_and_exists_amalgamation_iff_sheaf_for.2 q).1
+
+/-- C2.1.3 in Elephant -/
```

Comment:
Would you please make all these refs use the bib file? That way we get auto-hyperlinks in the docs.